core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArrayAggregationAfterDistinct.java

```diff
@@ -0,0 +1,162 @@
+                                getOnlyElement(functionCall.getArguments()) instanceof SymbolReference) {
+                            Expression functionArgument = getOnlyElement(functionCall.getArguments());
+                            Symbol aggregationSymbol = Symbol.from(functionArgument);
+                            if (aggregationSymbol.equals(outputSymbol)) {
+                                arrayDistinctSymbols.add(entry.getKey());
+                                continue;
+                            }
+                        }
+                    }
+                    if (SymbolsExtractor.extractUnique(expression).contains(outputSymbol)) {
+                        isReferenced = true;
+                        break;
+                    }
+                }
+
+                if (!arrayDistinctSymbols.isEmpty() && !isReferenced) {
+                    matchedAggregations.add(outputSymbol);
+                    aggregationsBuilder.put(outputSymbol, new Aggregation(
+                            aggregation.getResolvedFunction(),
+                            aggregation.getArguments(),
+                            true,
+                            aggregation.getFilter(),
+                            aggregation.getOrderingScheme(),
+                            aggregation.getMask()));
+                    for (Symbol symbol : arrayDistinctSymbols) {
+                        matchedAssignments.add(symbol);
+                        assignmentsBuilder.put(symbol, outputSymbol.toSymbolReference());
+                    }
+                }
+            }
+        }
+
+        if (!matchedAggregations.isEmpty()) {
+            // re-add not matched assignments
+            projectNode.getAssignments().entrySet().stream()
+                    .filter(entry -> !matchedAssignments.contains(entry.getKey()))
+                    .forEach(entry -> assignmentsBuilder.put(entry.getKey(), entry.getValue()));
+
+            // re-add not matched aggregations
+            aggregationNode.getAggregations().entrySet().stream()
+                    .filter(entry -> !matchedAggregations.contains(entry.getKey()))
+                    .forEach(entry -> aggregationsBuilder.put(entry.getKey(), entry.getValue()));
+
+            AggregationNode newAggregationNode = new AggregationNode(
+                    context.getIdAllocator().getNextId(),
+                    aggregationNode.getSource(),
+                    aggregationsBuilder.buildOrThrow(),
+                    aggregationNode.getGroupingSets(),
+                    aggregationNode.getPreGroupedSymbols(),
+                    aggregationNode.getStep(),
+                    aggregationNode.getHashSymbol(),
+                    aggregationNode.getGroupIdSymbol());
+            return Result.ofPlanNode(new ProjectNode(context.getIdAllocator().getNextId(), newAggregationNode, assignmentsBuilder.build()));
```

Comment:
I think this version does not contain the optimization, which the previous version had:
```
array_distinct(array_agg(DISTINCT x)) --> array_agg(DISTINCT x)
```
(do not apply `array_distinct` in the ProjectNode if the aggregation is `distinct` already).
Please add a test for this path too.